### PR TITLE
Elasticsearch Salt State

### PR DESCRIPTION
**Pull request by dextersd**

### CODE REVIEW

File: salt/elasticsearch/init.sls [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+
+elasticsearch:
+    # add repo to managed repositories
+    pkgrepo.managed:
+        - name: deb https://artifacts.elastic.co/packages/7.x/apt stable main
+        - key_url: salt://elasticsearch/release-key
+        - require:
+            - pkg: elasticsearch-deps
+    # install elastic search package
+    # TODO versioning
```
Comment (thusoy) [inline salt/elasticsearch/init.sls]: If we don't need anything more specific than 7.x atm this is probably fine and we can add more specifics later, so you can remove the TODO.

File: salt/elasticsearch/init.sls [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+        - require:
+            - pkgrepo: elasticsearch
+
+    # start service and watch the config files for restarting the service
+    service.running:
+        - watch:
+          - file: /etc/elasticsearch/elasticsearch.yml
+          - file: /etc/elasticsearch/jvm.options
+
+/etc/elasticsearch/jvm.options:
```
Comment (thusoy) [inline salt/elasticsearch/init.sls]: The target file name should be set in the `- name: <filename>` not the statement id, as this makes it very hard to customize the location or refer to it in another statement. Ie this should be `elasticsearch-jvm-options` or similar, and the statement below for the config could probably be put under the `elasticsearch` group (but you can only have one `file` statement per group thus you can't have both there).

File: salt/elasticsearch/jvm.options [hunk 3]
```diff
@@ -0,0 +1,79 @@
+{% set elasticsearch = pillar.get('elasticsearch', {}) -%}
+
+## JVM configuration
```
Comment (thusoy) [inline salt/elasticsearch/jvm.options]: Add a file marker here too.

File: salt/elasticsearch/map.jinja [hunk 4]
```diff
@@ -0,0 +1,7 @@
+{% set pillar_get = salt['pillar.get'] %}
+
+{% set elasticsearch = salt['grains.filter_by']({
+    'default': {
+        'version': '7.6'
```
Comment (thusoy) [inline salt/elasticsearch/map.jinja]: This isn't used, is it necessary or can we skip it?

File: pillar/test.sls [hunk 5]
```diff
@@ -2,6 +2,13 @@ sentry_forwarder:
     port: 5010
     sampling_rate: 10
 
+elasticsearch:
+    es_cluster_name: medal-test-cluster
+    es_memory: 1g
+    es_seed_hosts:
```
Comment (thusoy) [inline pillar/test.sls]: The keys are already namespaced under the `elasticsearch` key, so I don't think we need the `es_` prefixes in addition to that.

**Comment by dextersd** [comment]
@thusoy I've pinned us at 7.6.1 for the version, I think that makes more sense than 7.x.

As for how we handle configuration, I personally prefer templating the actual config file as you know what will be replaced by pillar and what is set by default. I don't see what benefit you get from serializing everything into pillar (other than the files being pre-compiled). Is it difficult to access pillar values? I guess being able to pull back all values at once rather than 1 at a time (if encrypted) wouldn't be bad. I think we should choose 1 way to do it and stick with it

File: salt/elasticsearch/init.sls [hunk 6]
```diff
@@ -5,12 +5,11 @@ elasticsearch-deps:
 elasticsearch:
     # add repo to managed repositories
     pkgrepo.managed:
-        - name: deb https://artifacts.elastic.co/packages/7.x/apt stable main
+        - name: deb https://artifacts.elastic.co/packages/7.6.1/apt stable main
```
Comment (thusoy) [inline salt/elasticsearch/init.sls]: I couldn't find a repo for this, to pin the version I think we need to specify the version to `pkg.installed` instead. Docs here: https://docs.saltstack.com/en/latest/ref/states/all/salt.states.pkg.html#salt.states.pkg.installed

**Comment by thusoy** [comment]
For the pillar_check to actually do anything you need to include it from the `init.sls`.

**Comment by thusoy** [comment]
I think the only thing missing here is opening up the firewall for port 9200 and 9300. There's an [example here](https://github.com/get-wrecked/salt-states/blob/master/salt/unattended-upgrades/init.sls#L33) that allows outbound DNS and multiple ports, we're doing something very similar where we want to allow outbound DNS and 9300 for the ES user, and inbound tcp 9200 and 9300.

File: salt/elasticsearch/init.sls [hunk 7]
```diff
@@ -36,3 +36,34 @@ elasticsearch-elasticsearch-yml:
         - name: /etc/elasticsearch/elasticsearch.yml
         - source: salt://elasticsearch/elasticsearch.yml
         - template: jinja
+
+
+{% set user = '_apt' if grains['os_family'] == 'Debian' and grains['osmajorrelease']|int >= 9 else 'root' %}
```
Comment (thusoy) [inline salt/elasticsearch/init.sls:41]: This can be removed, we're always using the `elasticsearch` user.

File: salt/elasticsearch/init.sls [hunk 8]
```diff
@@ -36,3 +36,34 @@ elasticsearch-elasticsearch-yml:
...
+        - table: filter
+        - chain: OUTPUT
+        - family: {{ family }}
+        - proto: tcp
+        - dport: 9300
+        - match:
+            - comment
+            - owner
+        - comment: 'elasticsearch: Allow outgoing traffic for for internal comms'
+        - uid-owner: {{ user }}
```
Comment (thusoy) [inline salt/elasticsearch/init.sls:55]: Harcode `elasticsearch` here and add a
```yaml
- require:
    - pkg: elasticsearch
```
to make sure the user exists before applying the rules. You should also include the DNS permissions as I'm planning to use that from the start.

File: salt/elasticsearch/init.sls [hunk 9]
```diff
@@ -36,3 +36,34 @@ elasticsearch-elasticsearch-yml:
...
+    firewall.append:
+        - table: filter
+        - chain: OUTPUT
+        - family: {{ family }}
+        - proto: tcp
+        - dport: 9300
+        - match:
+            - comment
+            - owner
+        - comment: 'elasticsearch: Allow outgoing traffic for for internal comms'
```
Comment (thusoy) [inline salt/elasticsearch/init.sls:54]: (minor) Typo: `for for`

**Pull request merged**